### DISCUSSION TITLE
Clearing a SourceList?

### DISCUSSION THREAD

**Discussion by houstonhaynes** [JordanMarr/ReactiveElmish.Avalonia#48] [Q&A] [answered]
@JordanMarr this is an odd question - as it's really a two-parter. So I sending a message from my MainViewModel to my ChatViewModel. I have a button (that resides in MainVM) which dispatches a command to clear the SourceList that populates the chat items in the chat window. 
![Screenshot 2024-01-12 at 9 48 46 PM](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/a8a84f78-c399-4b9c-b16b-21c10c0dce65)
Here's the ClearChatCommand in the ChatView.axaml
![Screenshot 2024-01-12 at 9 49 22 PM](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/1d5cd58b-8bbd-4dcb-8eb3-db03f90a0cd7)
And the binding in the ViewModel

```fsharp
member this.ClearChatCommand() =
        local.Dispatch ClearChatCommand
```
and the branch in the update function

```fsharp
| ClearChatCommand ->
            ChatViewModel.DesignVM.ClearChat()
            model
```

So far, so good, right? I know! I'm as surprised as anyone. 😸 

In the ChatViewModel here's ClearChat()

```fsharp
member this.ClearChat() =
        local.Dispatch(ClearChat)
```

Which then executes in the update function in the "ClearChat" branch\

```fsharp
| ClearChat ->
            printfn "clearing chat history"
            model.Messages.Clear()
            model
```
and I've *also* tried
```fsharp
| ClearChat ->
            printfn "clearing chat history"
            { model with Messages = model.Messages |> SourceList.removeAll }
```
and then - similar to the init()
```fsharp
| ClearChat ->
            printfn "clearing chat history"
            { model with Messages = SourceList.createFrom [] }
```
and finally in desparation I'm running the init() - which in context here is pretty tame
```fsharp
| ClearChat ->
            printfn "clearing chat history"
            init() |> ignore
            model
```

Even better - when I click the button - I see the console echo "clearing chat history" faithfully every time the action is taken. And yet, *nothing happens in the chat window*. All of the chat messages are still present. :/

Was it something I said, something I didn't say? Is there something about SourceList that I need to know. 

1. Am I routing the command from the MainView to the ChatView in a way that's consistent with your concept of passing messages from MVU loop to MVU loop? I *tried* to use a mechanism similar to what you used for "GoHome" but thought the "direct route" from Main to Chat view made sense here. Should I route through App?
2. Is there something I'm missing about either how SourceList works, or how returning the model with an initialized "Messages" should be asserted?

Any help/pointers/advice would be appreciated!

**Comment by JordanMarr** [comment]
Could it be related to using the Items property?

**Reply by houstonhaynes** [reply]
I'm not using Items here that I know of. (in fact I'm still not using it in the DataGrid in the other view because of that odd error)

**Reply by houstonhaynes** [reply]
Wait - I *am* using an ItemsSource reference - is that salient?
```fsharp
member this.MessagesView = this.BindSourceList(local.Model.Messages)
```

![Screenshot 2024-01-13 at 1 35 46 PM](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/5aecc600-2172-46f3-85cc-820fda93d0ae)

```XML
<DockPanel>
            <ListBox x:Name="ChatWindow" ItemsSource="{Binding MessagesView}"
                     CornerRadius="10"
                     ScrollViewer.BringIntoViewOnFocusChange="True"
                     ScrollViewer.AllowAutoHide="True"
                     ScrollViewer.VerticalScrollBarVisibility="Auto">
                <ListBox.ItemsPanel>
                    <ItemsPanelTemplate>
                        <StackPanel/>
                    </ItemsPanelTemplate>
                </ListBox.ItemsPanel>
                <ListBox.ItemTemplate>
                    <DataTemplate>
                        <StackPanel HorizontalAlignment="Stretch">
                            <StackPanel.Styles>
                                <Style Selector="Border">
                                    <Setter Property="Margin" Value="0, 5,50,5" />
                                    <Setter Property="BorderBrush" Value="{Binding BorderColor}" />
                                    <Setter Property="Background" />
                                </Style>
                                <Style Selector="Border.isMe">
                                    <Setter Property="Margin" Value="50,5,0,5" />
                                    <Setter Property="BorderBrush" Value="Gray" />
                                    <Setter Property="Background" />
                                </Style>							
                            </StackPanel.Styles>
                            <Border CornerRadius="10"
                                    Padding="10" 
                                    BorderThickness="2"
                                    Classes.isMe="{Binding IsMe}">
                                <StackPanel>
                                    <TextBlock FontWeight="Bold" Text="{Binding User}" Foreground="Tan" />
                                    <TextBlock TextWrapping="Wrap" Text="{Binding Text}" />
                                </StackPanel>
                            </Border>
                        </StackPanel>
                    </DataTemplate>
                </ListBox.ItemTemplate>
            </ListBox>
        </DockPanel>
```
![Screenshot 2024-01-13 at 1 34 25 PM](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/42db9394-15cc-4402-a86b-ecae4865b0c1)

**Reply by houstonhaynes** [reply]
The *dreaded* ReadOnlyObservableCollection. I didn't see it until now...

**Reply by JordanMarr** [reply]
SourceList.removeAll should work. 
And i assume you are using BindSourceList.

**Reply by JordanMarr** [reply]
What was the issue?

**Reply by houstonhaynes** [reply]
This? 

```fsharp
| ClearChat ->
            printfn "clearing chat history"
            { model with Messages = model.Messages |> SourceList.removeAll }
```

Doesn't work - also 

```fsharp
| ClearChat ->
            printfn "clearing chat history"
            model.Messages.RemoveRange(0, model.Messages.Count - 1 )
            model

```

Doesn't work

**Reply by JordanMarr** [reply]
But it does work the sample.

**Reply by houstonhaynes** [reply]
I don't know what to say - it's a conundrum

**Reply by houstonhaynes** [reply]
I just tried this method - same *non* result...

```fsharp
| ClearChat ->
      printfn "clearing chat history"
      let resetMessages = seq<ChatMessage>[]
      { model with Messages = model.Messages |> SourceList.removeAll |> SourceList.addRange resetMessages }
```

And for the record I'm on ReactiveElmish.Avalonia 1.0.2

**Comment by JordanMarr** [comment]
Hard to tell what the issue is. Your individual snippets all look fine.

**Comment by JordanMarr** [comment]
I just realized that you are calling the `DesignVM` property to clear the chat. 

```F#
        | ClearChatCommand ->
            ChatViewModel.DesignVM.ClearChat() // <- This is the problem
            model

```

**Reply by houstonhaynes** [reply]
I've tried so many ways to get this done I've lost my mooring.

**Comment by JordanMarr** [comment, marked as answer]
There are a bunch of ways you could call the `ClearChat()` method from the `MainViewModel`.
The easiest way would probably be to just do this:
```F#
    member this.ClearChatCommand() =
        let chatView = this.ChatView :> StyledElement
        let chatVM = chatView.DataContext :?> ChatViewModel
        chatVM.ClearChat()
```

There is no real value to be gained by pushing that into the local Elmish loop, so better to just delete the `ClearChatCommand` message in your `MainViewModule` and call the `ChatViewModel` directly from your `MainViewModel`.

Another way to do it would be to move your chat messages into the `App` module as global app state. But it's not worth breaking up the modularity you have just to facilitate sharing the ClearChat method.

Yet another way would be to use the EventBus pattern that I had in one of the earlier versions of the sample app.

**Reply by houstonhaynes** [reply]
Outstanding - thanks. We talked about moving chat into the app module but have held off. Your notes are really great food for thought. 👍
